Add doc for what issues Precaution catch (#242)
vmware/precaution: docs/issues_catched.md

# Issues Precaution finds

> Precaution uses language-specific security static code analysis tools (linters) to identify issues, therefore the amount and types of checks vary by programming language. Some of the types of issue which Precaution detects include:

Reviewer: ```suggestion
Precaution uses language-specific security static code analysis tools (linters) to identify issues, therefore the amount and types of checks vary by programming language. Some of the types of issues which Precaution detects include:
```

> Precaution uses language-specific security static code analysis tools (linters) to identify issues, therefore the amount and types of checks vary by programming language. Some of the classes of issue which Precaution detects include:

Reviewer: "classes of issue" - sounds strange.  "types of issues" sounds fine.

* [Hardcoded credentials](https://cwe.mitre.org/data/definitions/798.html)
* [SQL injection](https://cwe.mitre.org/data/definitions/89.html)
* [Command injection](https://cwe.mitre.org/data/definitions/77.html)
* [Weak random number generator](https://cwe.mitre.org/data/definitions/330.html)
* [Weak cryptographic algorithms](https://cwe.mitre.org/data/definitions/327.html)
* [Insecure libraries](https://cwe.mitre.org/data/definitions/829.html)
* [Specific insecure functions from libraries](https://cwe.mitre.org/data/definitions/749.html)
* [Inadequate encryption strength](https://cwe.mitre.org/data/definitions/326.html)
* [Eval function used](https://cwe.mitre.org/data/definitions/95.html)
* [TLS turned off or bad versions used](https://cwe.mitre.org/data/definitions/295.html)
* [Bad file permissions set during file creation](https://cwe.mitre.org/data/definitions/378.html)
* [Poor file permissions used with chmod](https://cwe.mitre.org/data/definitions/732.html)

**Note:** Precaution (through the underlying tools) warns developers of specific sections of code that may require security review.

> As with all automated detection tools, there will be cases of false positives.

Reviewer: rather than say definitely "there will be", I'd prefer the phrases to be "there is the possibility"

> As with all automated detection tools, there is a possibility there will be cases of false positives.

Reviewer: ```suggestion
As with all automated detection tools, there is a possibility of false positives.
```

# Issues Precaution finds

Precaution utilizes multiple language-specific security static code analysis tools (linters) 

> to identify issues for a particular language.

Reviewer: Redundant to state "language-specific" and "for a particular language"

Therefore the amount and types of checks for the different programming languages ​​differ.

> However, some of those rules can be grouped:

Reviewer: You mention "rules" here as if it was mentioned before, but it was not.

# Issues Precaution finds

Precaution utilizes multiple security static code analysis tools (linters)
to identify issues for a particular language.
Therefore the amount and types of checks for the different programming languages differ.

However, some of the rules used by the different linters can be grouped:

> 

Reviewer 2: This introduction feels a bit awkward, how about:

> Precaution uses language-specific security static code analysis tools (linters) to identify issues, therefore the amount and types of checks vary by programming language. Some of the classes of issue which Precaution detects include:

Author: Sounds a lot cleaner and easier to read.

* [Command injection](https://cwe.mitre.org/data/definitions/77.html)
* [Weak random number generator](https://cwe.mitre.org/data/definitions/330.html)
* [Weak cryptographic algorithms](https://cwe.mitre.org/data/definitions/327.html)
* [Insecure libraries](https://cwe.mitre.org/data/definitions/829.html)
* [Specific insecure functions from libraries](https://cwe.mitre.org/data/definitions/749.html)
* [Inadequate encryption strength](https://cwe.mitre.org/data/definitions/326.html)
* [Eval function used](https://cwe.mitre.org/data/definitions/95.html)
* [TLS turned off or bad versions used](https://cwe.mitre.org/data/definitions/295.html)
* [Bad file permissions set during file creation](https://cwe.mitre.org/data/definitions/378.html)
* [Poor file permissions used with chmod](https://cwe.mitre.org/data/definitions/732.html)

**Note:** Precaution (through the underlying tools) warns developers of specific sections of code that may require security review.
As with all automated detection tools, there will be cases of false positives.
You can read how to handle them here: [False positives and how to handle them](false_positivies.md)

> Full documentation for the supported security checks (rules)

Reviewer 2: Minor nit, a colon should precede a list
```suggestion
Full documentation for the supported security checks (rules):
```

> # Issues Precaution catches

Reviewer: ```suggestion
# Issues Precaution finds
```

Precaution utilizes multiple security static code analysis tools.

> Because of that the amount and types of checks for the different

Reviewer: Avoid starting a sentence with Because.

programming languages differs.
The security static code analyzers are language-focused
and some of the vulnerabilities they find are specific 

> for the programming language.

Reviewer 2: How about something a bit shorter like:

>Precaution utilizes multiple security static code analysis tools which are language-focused identify issues which may be specific to the language they target, therefore the amount and types of checks for the different programming languages differs.

> for the concrete programming language.

Reviewer: Why is concrete used here?

Author: It’s a misspelling but I will remove that word from here at all.

> However, some of those rules can be grouped into a few groups:

Reviewer: "grouped into few groups" sounds awkward.

* Hardcoded credentials

> * SQL injections

Reviewer: ```suggestion
* SQL injection
```

Author: I don’t understand. Above we are using “hardcoded credentials” but here you prefer that we use “SQL injection”?

> * Command injections

Reviewer: ```suggestion
* Command injection
```

> * HTML injections

Reviewer: ```suggestion
* HTML injection
```

Author: Same comment here as the one for the SQL and command injection.

* Weak random number generator

> * Imported and used weak cryptographic algorithms

Reviewer 2: This could be generalised to "Weak cryptographic algorithms"

> * Imported/used weak cryptographic algorithms

Reviewer: Why forward slash?

Author: Yes, you are right. If you import weak cryptographic algorithms you are going to use them too.

> * Imported insecure libraries

Reviewer 2: generalise to "Insecure libraries"

> * Use of specific insecure functions from libraries

Reviewer 2: "Specific insecure functions from libraries"

> * Weak key used for a cryptographic algorithm

Reviewer 2: Move under "Weak cryptographic algorithms" (to group like items), generalise to "Weak cryptographic keys" or, to be inline with CWE, "Inadequate encryption strength"

> * Eval function used

Reviewer 2: Is that all uses of `eval` or just uses which call `eval` on user-supplied data?

Author: In Bandit is whenever the function `eval` is used https://bandit.readthedocs.io/en/latest/blacklists/blacklist_calls.html#b307-eval
in TSLint-config-security is whenever `eval` is used with a non-literal argument: 
https://github.com/webschik/tslint-config-security#tsr-detect-eval-with-expression

> * Bind to all network interfaces example when using IP "0.0.0.0"

Reviewer: ```suggestion
* Binding to all network interfaces example when using IP "0.0.0.0"
```

This is an incomplete sentence.

> * SSL/TLS turned off or bad versions used

Reviewer: SSL is a very old term.  TLS is the proper name.

> * Set bad file permission when creating

Reviewer 2: This sentence is a little awkward, perhaps
```suggestion
* Bad file permissions set during file creation
```

> * Poor file permissions used with chmod

Reviewer: chmod is a command, not typically the exact API.  So I would avoid stating that.

Author: Yep, maybe you are right. That’s too specific.

> **Note:** Those vulnerabilities aims to warn the developers to audit a specific piece of code.

Reviewer 2: Perhaps something like:

```suggestion
**Note:** Precaution (through the underlying tools) warns developers of specific sections of code that may require security review.
```

Author: Yes, I don’t know why I wrote it the way I did :D. It’s confusing.

> **Note:** Those vulnerabilities aims to warn the developers to audit the concrete piece of code.

Reviewer: Sentence sounds awkward.  Why concrete?

Author: I will change the word but the idea is that we don’t want to warn the developers to audit the whole code base because then the linters are useless but for a specific code.

# Issues Precaution finds

Precaution utilizes multiple security static code analysis tools (linters)

> which are language-focused identify issues which may be specific to the language they target

Reviewer 2: this line isn't quite right:
"which are language-focused _and therefore_ identify issues which may be specific to the language they target"?

Author: Do you think that now is better?

Reviewer 2: LGTM
